zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/BackupStorageType.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.zookeeper.server.backup;
+
+/**
+ * Enum class that defines all implementations of BackupStorageProvider interface with corresponding
+ * class names.
+ */
+public enum BackupStorageType {
```

Comment:
This seemd to fixed type for supporting. Are we planning to have it more flexible? For example, server can load the runtime defined class path anytime.

Reply:
I'm not sure if that's a use case we want to / need to support. Could you explain why that would be needed? I think that would complicate things with little benefit.

Reply:
We can make it as generic support. Saying on cloud environment, we can let user to define their own backup storage (remote backups) and logic instead of changing Zookeeper server code.

If we would like to support general API for backup storage, it is better we leave the flexibility of implementation. Otherwise, the API definition will be constrained by these fixed types.

Reply:
Thank you for the comment. Can you explain a little more about what you exactly mean by generic support?

I think we might be thinking about two different things:
1. the flexibility of users providing their own storage implementation (say, not tied to hdfs or nfs, they should be able to provide a blob storage implementation if they wanted to)
2. the feature that can dynamically switch and load backup implementation classes on the fly while the backup node is running 

Is 2 what you're referring to? I don't see a strong need/use case for 2. If you are referring to 1, we already accomplish this - users can freely add their own implementation to the server code base since the ZK server code base is open source and is available. Although not shown in this PR, the server will instantiate the appropriate storage backend at runtime based on the string value provided in JVM properties using reflection. That's why each enum defined is required to have a fully-qualified class name for its corresponding implementation.

Reply:
I think @dasahcc is referring to 1 and he was asking if there a way we can let user input an enum that they define so they don't need to change zk server code?

Reply:
Thanks @narendly and @mgao0 . That's my point. On fly upgrade is not necessary. But the enum is still ZK code. User can add their code in a separate package / class for new things. Change enum class, from logic perspective, is something user should not do. That's why I do suggest to make the type more flexible in definition.

Reply:
@dasahcc Users do not have to modify the enum class. If they add a storage implementation, they only need to supply the fully-qualified name of the class in the JVM Properties.

The enum is added to indicate which classes are officially supported as part of the ZK codebase. But since we aren't even sure if this will be merged into upstream, I'll just remove this enum class so we could get past this.